### PR TITLE
test(iframes): Add cross-origin clicks (#3170)

### PR DESCRIPTION
**Pull request by rwoll**
This PR adds a variety of tests regarding clicking on links in iframes. Based on the numerous edge cases surfaced by these tests, I think it would be good to add more iframe click tests in some form or another. (These tests were authored in real-time as I was discovering more edge cases, so I'm very open to suggestions for improvement. ❤️ )

EDIT: With #3201 by @dgozman, the Chrome Headfull tests are now passing! 🎉 

~~Many of the tests mark Chromium Head**full** as failing. In several of the failing Chromium Headfull failing cases, Playwright reports a click (i.e. we get past an `await button.click()`), but the click doesn't seem to actually happen (See `should click the visible link in an iframe (cross-origin)`.) In other cases, we timeout waiting for the click to complete and observe the following (repeated) log:~~

```
  pw:api retrying click action [] +0ms
  pw:api   waiting for element to be visible, enabled and not moving [] +0ms
  pw:api   element is visible, enabled and does not move [] +26ms
  pw:api   scrolling into view if needed [] +0ms
  pw:api   done scrolling [] +0ms
  pw:api   checking that element receives pointer events at (35.32,17) [] +1ms
  pw:api   element does not receive pointer events [] +6ms
```

~~See `should click the visible link in an iframe (cross-origin) in fixed position div` for an example of this happening.~~

Finally, we get to the set of tests that deal with [`X-Frame-Options`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options) which is set to `DENY` to match the Wikipedia login link:

```
$ curl -v 'https://en.wikipedia.org/w/index.php?title=Special:UserLogin&returnto=Main+Page' 2>&1 | grep -i x-frame-options
< x-frame-options: DENY
```

In these tests—once we account for the clicking issues mentioned above—we see inconsistent behavior. Assuming you run the following script:

```javascript
const http = require("http");
const server = http.createServer((req, res) => {
  res.setHeader("Cache-Control", "no-store");
  console.log(req.url);
  res.end(`
    <body>
        <div>
            <iframe src="https://en.wikipedia.org/"  width="100%" height="100%" > </iframe>
        </div>
    </body>
  `);
});

server.listen(3000);
```
If you click _login_ in the iframe, the consumer-facing versions of Chrome, Firefox, and Webkit will all block the action and warn you in the console that `X-Frame-Options` is set to `'deny'`. However, in the Playwright versions of the browsers (as you'll see by the browser-specific passing criteria for the `should click and navigate to a x-frame-options:DENY link` test, Webkit and Chromium (Headless) give you the deny error, while Firefox navigates to the page against the policy set in the header. (Should Playwright-Firefox match the consumer Firefox behavior?)

### CODE REVIEW

**Comment by rwoll** [comment]
I've added two more tests in 8142170 which more closely match the reported scenanrio:

This changeset adds tests the more closely match the reported scenario
in #3170. Firefox, both headless and headfull pass completetly in all
cases. The other browsers (both headless and headfull) report a
successful click (i.e. they get past the await `button.click()`) but
fail to pass the navigation check, except for Chromium HeadFULL with a
fixed div which fails to even do the click.

NB: If you perform this test manually in the production version of
Firefox (78.0.2), the navigation to the Wikipedia login page will be
blocked due to X-Frame-Options: DENY. The iframe will load on
localhost, but clicking login will get you a an error about
X-Frame-Options. So, in some ways, even though this test is "passing"
for FFOX, in a traditional user environment we'd expect it to fail.

**Comment by rwoll** [comment]
Added _WIP_ until I clean up the failing test syntax and also assert against the console messages that say:

```
Refused to display 'http://127.0.0.1:8907/login-with-x-frame-options-deny.html' in a frame because it set 'X-Frame-Options' to 'DENY'.
```

which should be considered passing.

File: test/click.jest.js [hunk 1]
```diff
@@ -307,6 +307,177 @@ describe('Page.click', function() {
...
+    const button = await frame.$('#pt-login');
+    await button.click();
+
+    if (FFOX) {
+      expect(await navigated).toBeTruthy();
+    } else if (WEBKIT || CHROMIUM && HEADLESS) {
+      expect((await consoleMessaged).match(/^Refused to display.*login-with-x-frame-options-deny\.html' in a frame because it set 'X-Frame-Options' to 'DENY'\./i)).toBeTruthy();
+    } else if (CHROMIUM && !HEADLESS) {
+      expect((await consoleMessaged).match(/^Failed to load resource: the server responded with a status of 404 \(Not Found\)/i)).toBeTruthy();
+    }
```
Comment (rwoll) [inline test/click.jest.js:479]: @aslushnikov What should I baseline cases against? Consumer version of Firefox would fall into the  `Refused to display` category, but in Playwright it allows the frame navigation.

**Comment by rwoll** [comment]
I've updated the description to give a better overview of the current state of the MR and where tests fail or diverge from expectation.

@aslushnikov Can you add this to your review queue and let know if the description and the tests themselves make sense? The PR ended up a bit larger than I anticipated, so hoping it's still followable.

File: test/click.jest.js [hunk 2]
```diff
@@ -307,6 +307,193 @@ describe('Page.click', function() {
...
+        fulfill(msg.text());
+      });
+    });
+
+    const frame = page.frames()[1];
+    const button = await frame.$('#pt-login');
+    await button.click();
+
+    if (FFOX) {
+      expect(await navigated).toBeTruthy();
```
Comment (rwoll) [inline test/click.jest.js:436]: Consumer Firefox would actually block this navigation; however, Playwright-Firefox does not block it. Should Playwright-Firefox block it?

Comment (dgozman) [reply test/click.jest.js:436]: I think this is a bug. We should mark the test as fail(FFOX).

**Review by dgozman — COMMENTED** [review]
Thank you for the PR!

We usually prefer smaller more focused tests, because those are easier to debug if anything goes wrong. For example, a test that has an iframe with `X-Frame-Options: DENY` and verifies that we do not display it is a great one. Would you mind to focus on that specific aspect in this PR?

I think that with the recent fix in Chromium, various click tests do not provide as much value, because every aspect is being tested separately: clicking in an cross-process vs same-process iframe, clicking in a fixed position div, etc.

Overall, this is some impressive work!

**Comment by rwoll** [comment]
Thanks for the comments, and agree many of these tests no longer provide value given your more targeted tests (and lovely fix)! I'm going to close this MR and open one tightly scoped around x-frame-options. Since the clicking is sorted out, I'll try to avoid conflating the two concerns in the new tests.
